src/async/pitfalls/async-traits.md

```diff
@@ -0,0 +1,50 @@
+# Async Traits
+
+Async methods in traits are not yet supported in the stable channel ([An experimental feature exists in nightly and should be stabilized in the mid term.](https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html))
+
+The crate [async_trait](https://docs.rs/async-trait/latest/async_trait/) provides a workaround through a macro:
+
+```rust,editable,compile_fail
+use async_trait::async_trait;
+use tokio::time::{sleep, Duration};
+
+#[async_trait]
+trait Sleeper {
+    async fn sleep(&self);
+}
+
+struct FixedSleeper {
+    sleep_ms: u64,
+}
+
+#[async_trait]
+impl Sleeper for FixedSleeper {
+    async fn sleep(&self) {
+        sleep(Duration::from_millis(self.sleep_ms)).await;
+    }
+}
+
+async fn run_all_sleepers_multiple_times(sleepers: Vec<Box<dyn Sleeper>>, n_times: usize) {
+    for _ in 0..n_times {
+        for sleeper in &sleepers {
+            sleeper.sleep().await;
+        }
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let sleepers: Vec<Box<dyn Sleeper>> = vec![
+        Box::new(FixedSleeper { sleep_ms: 50 }),
+        Box::new(FixedSleeper { sleep_ms: 100 }),
+    ];
+    run_all_sleepers_multiple_times(sleepers, 5).await;
+}
+```
+
+<details>  
+
```

Comment:
Perhaps add a note on why this is not supported (i.e., the resulting type not having a known type at compile time), and the overhead of using async_trait (because it does heap allocation).

Reply:
By "supported", do you mean why this isn't stabilized?

Reply:
Yes :)

Reply:
Sure. I can add something once you've merged #556. I think I learnt about it in an episode of Crust of Rust. We could also leave it out, as it might be too much detail. 

I think the instructors can choose to go into more depth as they see fit when teaching the course. We don't need to cover everything in the notes, but since the speaker notes for this section are are already so comprehensive, I was tempted to suggest adding this too :)

Reply:
If think I know why but I would have a hard time explaining, so let me provide a pointer and you can explain better than me:

If you had a trait with `fn fetch_data_count(&self) -> usize`, then the compiler knows that it needs to push 64 bits to the stack for the return value, whatever the type of the caller.

Now if we had `async fn fetch_data_count(&self) -> usize`, since we actually return a future, the stack of the future needs to also be allocated, so the size of the return value depends on the actual method.

Consider both implementations:
```
async fn fetch_data_count(&self) -> usize {
    let data: [i32; 1000] = fetch().await;
    let filter: i32 = fetch_filter().await;
    data.iter().filter(|&item| item != filter).count()
}
```

```
async fn fetch_data_count(&self) -> usize {
    let data: Vec<i32> = fetch().await;
    let filter: i32 = fetch_filter().await;
    data.iter().filter(|&item| item != filter).count()
}
```

Example one, with the data on the stack probably has a higher size than the second one with the Vec. Using dynamic dispatch, it would be impossible for the compiler to know ahead of time the size.

Reply:
Heap allocating, fixes the size issue. Note that, Box does not always allocate and the compiler is often able to omit allocations if it can figure out the size

Reply:
Yes. Exactly. And I just noticed that I have a typo in my original comment; it should be "the resulting type not having a known **size** at compile time". 

We cannot add all the details in the notes, so I was thinking of just a two sentence summary. I'll create a PR.